frontend/catalyst/compiler.py

```diff
@@ -155,6 +156,55 @@ def run_writing_command(command: List[str], compile_options: Optional[CompileOpt
         "adjoint-lowering",
     ],
 )
+QUANTUM_COMPILATION_PASS_deafult_len = len(QUANTUM_COMPILATION_PASS[1])
+
+"""
+# We would like the quantum passes pipeline to look something like the following:
+QUANTUM_COMPILATION_PASS = (
+    "QuantumCompilationPass",
+    [
+        # Begin quantum peephole
+        "func.func(remove-chained-self-inverse{func-name=circuit})",
+        "func.func(remove-chained-self-inverse{func-name=toy})",
+        ...
+        # End quantum peephole
+        "annotate-function",
+        ...
+    ],
+)
+
+The peephole circuit transforms should occur before all the lowering
+because the circuit transforms are independent of the lowering and do 
+not need to parse through all the lowered MLIR operations/instructions. 
+"""
+
+
+def fill_quantum_peephole_passes_pipeline(pass_table):
+    """
+    Fill the QUANTUM_COMPILATION_PASS pipeline according to the
+    pass_table. The pass table contains the list of passes to run
+    on each qnode.
+    """
+    quantum_passes_pipeline = []
+    for qnode in pass_table.table.keys():
+        for pass_ in pass_table.query(qnode):
+            # note that "pass" is a python keyword, so "pass_"
+            # The command line option to specify the qnode name is "func-name".
+            # See mlir/include/Quantum/Transforms/Passes.td
+            run_pass = f"func.func({pass_}{{func-name={qnode.__name__}}})"
+            if (run_pass not in quantum_passes_pipeline) and (
+                run_pass not in QUANTUM_COMPILATION_PASS[1]
+            ):
+                quantum_passes_pipeline.append(run_pass)
+    QUANTUM_COMPILATION_PASS[1][:0] = quantum_passes_pipeline
+
+
+def clear_quantum_peephole_passes_pipeline():
+    """
+    Clear the peephole passes in the QUANTUM_COMPILATION_PASS pipeline.
+    """
+    del QUANTUM_COMPILATION_PASS[1][: -1 * QUANTUM_COMPILATION_PASS_deafult_len]
+
```

Comment:
these two functions are internal and not user facing. Should there really be tests for them? 

Also, any tests on them would change the `QUANTUM_COMPILATION_PASS` pipeline global in the file.

Reply:
Non-user facing code should also be tested.